src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,226 @@
+ */
+final class ObjectMapper implements ObjectMapperInterface
+{
+    public function __construct(
+        private readonly MapperMetadataFactoryInterface $metadataFactory = new ReflectionMapperMetadataFactory(),
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+        private readonly ?ContainerInterface $serviceLocator = null
+    ) {
+    }
+
+    public function map(object $object, object|string $to = null): object
+    {
+        static $objectMap = null;
+        $objectMapInitialized = false;
+
+        if (null === $objectMap) {
+            $objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $refl = new \ReflectionClass($object);
+
+        $metadata = $this->metadataFactory->create($object);
+        $map = $this->getMapTo($metadata, null, $object);
+        $to ??= $map?->to;
+
+        if (!(\is_string($to) && class_exists($to)) && !\is_object($to)) {
+            throw new RuntimeException(sprintf('Attribute of type "%s" expected on "%s.', Map::class, $refl->getName()));
+        }
+
+        $toRefl = new \ReflectionClass($to);
+        $mapped = \is_object($to) ? $to : $toRefl->newInstanceWithoutConstructor();
+        if ($map && $map->transform) {
+            $mapped = $this->applyTransforms($map, $mapped, $mapped);
+        }
+
+        if (!\is_object($mapped)) {
+            throw new RuntimeException('Can not map to a non-object.');
+        }
+
+        $objectMap[$object] = $mapped;
+
+        $arguments = [];
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            $property = $toRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mapped) && $property->getValue($mapped))) {
+                continue;
+            }
+
+            $arguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
+            if ($property->isStatic()) {
+                continue;
+            }
+
+            $propertyName = $property->getName();
+
+            $map = null;
+            foreach ($property->getAttributes(Map::class) as $attr) {
+                $map = $attr->newInstance();
+                $if = $map->if;
+
+                if (false === $if) {
+                    continue 2;
+                }
+
+                if ($if && ($fn = $this->getCallable($if)) && !$this->call($fn, null, $object)) {
+                    continue 2;
+                }
+
+                break;
+            }
+
+            $mapToProperty = $map?->to ?? $propertyName;
+            if (!$mapToProperty || !$toRefl->hasProperty($mapToProperty)) {
```

Comment:
Certainly for latter: but i think that relations between source and target should be computed before doing the mapping : 

 * it allows this data to be cached (since it will only change if the source and/or target class change)
 * it allow to build debug tools on top on that to see what the mapper will do (command or profiler) and in this case debug is really interesting because it can be a "silent" error (typo in target property name)

Reply:
it would add another loop though, let's keep this as a future improvement

Reply:
metadata should compute the source => target thing this is a followup on my last comment.